trials at a fintech (#1)
fkato1/blog-usa: Trials-at-a-Fintech.md

---
authors:
- Francis Kato
tags:
date: 2021-10-20T10:00:00.000Z
title: "Trials at a Fintech"
---

> Fintechs have become a new frontier where innovation in the financial sector has been growing. Applications like credit and blockchain force companies in this sector to move quickly and provide services to new customers or pry them away from their competitors. Recently, I got the opportunity to consult at a well funded fintech company that provides financial services like access to fair and clear credit. These services are aimed at people who seek to build or rebuild their credit. This company which we shall call “Credit of all credit” is a newly founded stand alone credit card company. (https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2021/10/creditcard.jpeg)It’s umbrella company, which we shall call “Bank of all banks”, created it to serve as it’s credit card division. Based on their funding, they are a growing company with new clients being added regularly. This means they have a number of great opportunities around infrastructure, development/engineering, and product management. They mainly use new technologies in their projects which allows for learning new things and growth at the same time in various disciplines like microservices, cloud computing, and data engineering.

Reviewer: ```suggestion
Fintechs have become a new frontier where innovation in the financial sector has been growing. Applications like credit and blockchain force companies in this sector to move quickly and provide services to new customers in order to pry them away from their competitors.

Recently, I got the opportunity to consult at a well-funded fintech company that provides financial services, including access to fair and clear credit. These services are aimed at people who seek to build or rebuild their credit. This company, which we shall call “Credit of all credit,” is a newly-founded standalone credit card company. (https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2021/10/creditcard.jpeg) It’s umbrella company, which we shall call “Bank of all banks,” founded it to serve as its credit card division. Based on their funding, they are a growing company with new clients being added regularly. This means they have a number of great opportunities around infrastructure, development/engineering, and product management. They mainly use new technologies in their projects to allows for learning new things and growth at the same time in various disciplines like microservices, cloud computing, and data engineering.
```

> Operation Sour Patch Kids

Reviewer 2: Is Sour Patch Kids the ML for this operation? If yes, we shouldn't use it here.

Reviewer 3: It's umbrella company -> Its

(https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2021/10/sourpatch.jpeg)

> Operation Sour Patch Kids was a conversion from one core processor to another. The “previous” is a provider of technology solutions focused on serving merchants, enabling them to accept electronic payments, including card-based payments, contactless card and mobile wallet. They also serve financial institutions with core processing software, transaction processing software and applications and services. The “current” is a payment solutions provider whose products and services enable businesses of any size, and in any industry, to accept credit card payments in every way. The “current” is a world wide payment processor that handles billions of transactions a year. It has provided “Credit of all credit”‘ an excellent option of being a one-stop shop for all their credit card processing needs. This fintech also received its own, unique merchant account, rather than the aggregate account that many payment processors like the “previous” offer their customers. With the “current” there’s no need to work with a third-party merchant account provider. It will also be able to accept credit card payments via several points-of-sale, including in-person, online and on the go. The migration was meant to achieve the following objectives:

Reviewer: ```suggestion
Operation Sour Patch Kids was a conversion from one core processor to another. The “previous” is a provider of technology solutions focused on serving merchants, enabling them to accept electronic payments, including card-based payments that can be contactless / mobile wallet. They also serve financial institutions with core processing software, transaction processing software, and applications and services. The “current” is a payment solutions provider whose products and services enable businesses of any size, and in any industry, to accept credit card payments in many ways. The “current” is a worldwide payment processor that handles billions of transactions a year. It has provided “Credit of all credit”‘ an excellent option of being a one-stop shop for all their credit card processing needs. 

This fintech also received its own unique merchant account, rather than the aggregate account that many payment processors like the “previous” offer their customers. With the “current,” there is no need to work with a third-party merchant account provider. It will also be able to accept credit card payments via several points-of-sale, including in-person, online, and on-the-go. The migration was meant to achieve the following objectives:
```

Scale to support current and future card programs

> Allow for flexibility and allow for evolution

Reviewer 2: I would reword this. Maybe: `Allow for the platform to evolve with new features`

Affordability for all services provided
Extensibility to provide service beyond accounting
Compliance and risk management

Communications Platform

> Team BABL which included Adam Moeller, Davis Mohar, Kevin Dinh, Nathan Shurte, Jake Henningsgaard and myself were tasked with building a message orchestration service to handle all kinds of communication events by all domains. I joined the team as a new Ipponite in April when they were already three months into the project. The main functions of the platform:

Reviewer 2: Don't use Team BABL. Just say our team

Reviewer 2: I would leave out the names of the other consultants

> Team BABL which included Adam Moeller, Davis Mohar, Kevin Dinh, Nathan Shurte, Jake Henningsgaard and myself were tasked with building a message orchestration service to handle all kinds of communication events by all domains. I joined the team as a new Ipponite in April when they were already three months into the project. The main functions of the platform:

Reviewer: ```suggestion
Our team was tasked to build a message orchestration service to handle all kinds of communication events by all domains. I joined the team in April when they were already three months into the project. The main functions of the platform included:
```

> Make sure customer communication is central in one system

Reviewer: ```suggestion
Ensure customer communication is centralized to one system
```

Analyzing customer data

> Handle Tsys notifications

Reviewer 2: Tsys -> TSYS

> Handle Tsys notifications

Reviewer: ```suggestion
Handle TSYS notifications
```

> Ippon to the rescue

Reviewer: ```suggestion
Ippon to the Rescue
```

(https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2021/10/ippon.jpeg)

> The platform utilized GCP pub/sub for message communication amongst the microservices. The client had an internal streaming platform which was over pubsub and mainly as a security layer to validate payloads. This in house product allows for service to service communication based on async messages. The microservices were all written in kotlin and integration system testing was primarily done with Karate. Some of the other technologies used were AWS S3 for storage, Kubernetes, Docker for containerization, Postgres for databases, monitoring tools like Kibana logs, Grafana, and Pagerduty. They also utilized an impressive inbuilt monitoring service for their kubernetes clusters. Naturally, there was a huge learning curve with all these technologies but the foundation of the team certainly allowed for growth even though we were all using them for the first time. For each of these technologies, we had learning dojos to familiarize ourselves and try to uncover ways to improve usability in our own platform. Several of these were with GCP like message queuing with pub/sub where by we had a scenario that required a list of tasks to be executed, publisher sent messages/payloads to the subscriber. Another example we triggered a cloud function with a pub/sub event. This helped us learn better and gave us the opportunity to create small fun projects from which every team member was able to benefit from. As a team we also each contributed to each service whether it was adding a new feature or increasing code coverage and as a result knowledge was able to be shared evenly.

Reviewer 2: kotlin -> Kotlin

> The platform utilized GCP pub/sub for message communication amongst the microservices. The client had an internal streaming platform which was over pubsub and mainly as a security layer to validate payloads. This in house product allows for service to service communication based on async messages. The microservices were all written in kotlin and integration system testing was primarily done with Karate. Some of the other technologies used were AWS S3 for storage, Kubernetes, Docker for containerization, Postgres for databases, monitoring tools like Kibana logs, Grafana, and Pagerduty. They also utilized an impressive inbuilt monitoring service for their kubernetes clusters. Naturally, there was a huge learning curve with all these technologies but the foundation of the team certainly allowed for growth even though we were all using them for the first time. For each of these technologies, we had learning dojos to familiarize ourselves and try to uncover ways to improve usability in our own platform. Several of these were with GCP like message queuing with pub/sub where by we had a scenario that required a list of tasks to be executed, publisher sent messages/payloads to the subscriber. Another example we triggered a cloud function with a pub/sub event. This helped us learn better and gave us the opportunity to create small fun projects from which every team member was able to benefit from. As a team we also each contributed to each service whether it was adding a new feature or increasing code coverage and as a result knowledge was able to be shared evenly.

Reviewer: ```suggestion
The platform utilized GCP Pub/Sub for message communication amongst the microservices. The client had an internal streaming platform which was over pubsub and mainly as a security layer to validate payloads. This in-house product allows for service-to-service communication based on async messages. The microservices were all written in Kotlin and integration system testing was primarily done with Karate. Some of the other technologies used were AWS S3 for storage, Kubernetes, Docker for containerization, PostgreSQL as the databases, and monitoring and logging done by Kibana, Grafana, and PagerDuty. They also utilized an impressive in-house built monitoring service for their Kubernetes clusters.

Naturally, there was a huge learning curve with all these technologies, but the foundation of the team certainly allowed for growth, despite us all using them for the first time. For each of these technologies, we had learning dojos to familiarize ourselves and try to uncover ways to improve usability in our own platform. Several of these were with GCP, such as message queuing with pub/sub, whereby we had a scenario that required a list of tasks to be executed, followed by the publisher sending messages/payloads to the subscriber.

Another example we triggered was a cloud function with a pub/sub event. This helped us better learn and gave us the opportunity to create small, fun projects from which every team member was able to benefit from. As a team, we each contributed to each service whether it was adding a new feature or increasing code coverage. As a result, knowledge share became easier.
```

> During my time there I worked on all the services that were deployed. Whether it was improving an existing feature or adding a new one, understanding how events move through the platform helped you see the big picture. I also helped improve code coverage in all the services and touched on system integration testing with Karate. The team mapped out over 170 scenarios to thoroughly test the functionality of the platform. The main piece of work I did was with an api orchestrator. The main purpose is for frontends to interact with the backends or services that handle customer data. It utilized graphQL as the api query language and the data could only be requested from one endpoint. Our records service which stored all customer events directly interacted with the frontend the client’’s agents use. The main reason was to ease integration with an authentication service. This was my first time seeing this type of security layer. Initially it didn’t make sense because it seemed easier to directly hook up the endpoints with the frontend. However, because they handle finances ensuring that only the requested data in its proper format is shared protects the client and their customers. 

Reviewer: ```suggestion
During my time there, I worked on all the services that were deployed. Whether it was improving an existing feature or adding a new one, understanding how events move through the platform helped you see the big picture. I also helped improve code coverage in all the services and touched on system integration testing with Karate. The team mapped out over 170 scenarios to thoroughly test the functionality of the platform. The main piece of work I did was with an API orchestrator. The main purpose is for the front end to interact with the back end or services that handle customer data. It utilized GraphQL as the API query language and the data could only be requested from one endpoint. Our records service, which stored all customer events, directly interacted with the front end the used by our client's agents. The main reason was to ease integration with an authentication service.

This was my first time seeing this type of security layer. It initially did not make sense because it seemed easier to directly hook up the endpoints with the front end. However, because they handle finances, ensuring that _only_ the requested data in its proper format is shared protects both the client and their customers.
```

Outcome 
(https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2021/10/rickmorty.jpeg)

> The client had two dates in September for transition, soft launch and cutover. Cutover was the day when the new processor was in full effect. Ippon successfully managed to implement, develop and deliver the communications platform. During cutover weekend we experienced issues which increased our heartbeats and had us all in a small panic phase. For example, some of the payloads moving through our system were missing data. They were missing personal fields like names and IDs were either nonexistent or didn’t match. To mitigate this we stopped the flow of customer data to reduce the size of data that had to be rectified. We ran a comparison between our compromised data and what it should be to find the bad payloads. Also, one of the third party technologies that interacted with customer data didn’t trigger communication events that were expected. Along with the client’s engineers, we were able to move over those hurdles and spend more time making improvements around our communication events. I can only speak for our team and the platform we were mainly focused on. The client was able to get through to the finish line in decent shape. Some of the things they would have liked to do had to be pushed to a later date because they mainly ran out of time. All in all they were proud of the work that was accomplished on all the teams and the shareholders acknowledged that. Currently, finishing up tech debts and making improvements is priority but they are fully in production and customers are using the new platform.

Reviewer 2: I wouldn't be that specific. Just say a soft launch and cutover

> The client had two dates in September for transition, soft launch and cutover. Cutover was the day when the new processor was in full effect. Ippon successfully managed to implement, develop and deliver the communications platform. During cutover weekend we experienced issues which increased our heartbeats and had us all in a small panic phase. For example, some of the payloads moving through our system were missing data. They were missing personal fields like names and IDs were either nonexistent or didn’t match. To mitigate this we stopped the flow of customer data to reduce the size of data that had to be rectified. We ran a comparison between our compromised data and what it should be to find the bad payloads. Also, one of the third party technologies that interacted with customer data didn’t trigger communication events that were expected. Along with the client’s engineers, we were able to move over those hurdles and spend more time making improvements around our communication events. I can only speak for our team and the platform we were mainly focused on. The client was able to get through to the finish line in decent shape. Some of the things they would have liked to do had to be pushed to a later date because they mainly ran out of time. All in all they were proud of the work that was accomplished on all the teams and the shareholders acknowledged that. Currently, finishing up tech debts and making improvements is priority but they are fully in production and customers are using the new platform.

Reviewer: ```suggestion
The client had two dates in September for transition: **soft launch** and **cutover**.

Cutover was the day when the new processor was in full effect. Ippon successfully managed to implement, develop, and deliver the communications platform. During cutover weekend, we experienced issues which increased our heartbeats and had us all in a small panic phase. For example, some of the payloads moving through our system were missing data. They were missing personal fields like names and some IDs were either nonexistent or did not match. To mitigate this, we stopped the flow of customer data to reduce the size of data that had to be rectified. We ran a comparison between our compromised data and what it should be to find the bad payloads. In addition, one of the third party technologies that interacted with customer data did not trigger communication events that were expected.

Along with the client’s engineers, we were able to move over those hurdles and spend more time making improvements around our communication events. Of course, I can only speak for our team and the platform we were mainly focused on.

The client was able to get through to the finish line in decent shape. Some of the things they would have liked to do had to be pushed to a later date because they ran out of time. All in all, they were proud of the work that was accomplished on all the teams and the shareholders acknowledged that. Currently, finishing up tech debts and making improvements has become priority, but they are fully in production and customers are using the new platform.
```

Culture at “Credit of all credit”

> The client has a great culture because of the talent, resources, and projects they do. With enterprise clients, they'll build their own solutions for certain problems, and integrating with these solutions is sometimes easy and sometimes a huge pain. One thing we can do to "leave it better than we found it" is to contribute back to documentation on internal solutions. The engineers responsible for these projects are extremely knowledgeable and give a completely different perspective to development. The level of creativity amongst the teams also makes their projects engaging and fun to work on. The client is a hard working organization that tackles gigantic problems usually in a short period of time which is very impressive. The major con of this is that the quality of work will suffer even though the product delivers. The workload brought in a ton of new hires and that made working with the teams a little bit more difficult. This was a huge problem when it came to their existing in-house projects because finding an engineer to work you through it was hard. I think this is an area they can improve in. Their onboarding program is great and can get you up to speed quickly with all their technologies but sometimes having an engineer available can eliminate a lot of questions. One thing I realized was if you got to work with someone helpful then keep the contact and build a strong relationship with them.

Reviewer: ```suggestion
The client has a great culture because of the talent, resources, and projects they tackle. With enterprise clients, they will build their own solutions for certain problems, and integrating with these solutions can be easy or a huge pain. One thing we can do to "leave it better than we found it" is to contribute back to documentation on internal solutions. The engineers responsible for these projects are extremely knowledgeable and give a completely different perspective to development. The level of creativity amongst the teams also makes their projects engaging and fun to work on.

The client is a hardworking organization that solves gigantic problems in a short period of time. The major con of this is that quality of work can sometimes suffer, even if the product delivers. The workload brought in a ton of new hires and that made working with the teams a bit more difficult. This became a problem when it came to their existing in-house projects because finding an engineer with existing domain knowledge to pair with was difficult. This is an area in which improvement may prove valuable. 

Their onboarding process is efficient and can get you up to speed quickly with all their technologies. As hinted at before, however, sometimes having an engineer available can eliminate a lot of questions along the journey. One thing I realized was if you got to work with someone helpful, it was vital to keep contact and build a strong relationship with that individual.
```

Conclusion

> If they decide to tackle more enormous difficult projects in the future, then managing the workload and having a readily available workforce can produce a higher quality of work. 

Reviewer: ```suggestion
Should the client decide to take on more surmountable, difficult projects in the future, then managing the workload and having a readily available workforce can produce a higher quality of work.
```

> I definitely enjoyed working with the client mainly because of the people I got to team up with. Our Babl team is the best I have been on in my young career. It’s in a perfect industry that is evolving and can’t wait to see what new products they come up with. Now that they have managed to migrate successfully, I think it is now in a position to make strides forward.

Reviewer 2: remove Babl

> I definitely enjoyed working with the client mainly because of the people I got to team up with. Our Babl team is the best I have been on in my young career. It’s in a perfect industry that is evolving and can’t wait to see what new products they come up with. Now that they have managed to migrate successfully, I think it is now in a position to make strides forward.

Reviewer: ```suggestion
I definitely enjoyed working with the client, mainly because of the people I got to team up with. Our unit was by far the best I have been on in my young career. We were in a perfect industry that is evolving and I cannot wait to see what new products they come up with. Now that they have managed to migrate successfully, I think it is now in a position to make strides forward.
```
